perception/autoware_tensorrt_rtmdet/src/preprocess.cu

```diff
@@ -0,0 +1,460 @@
+// Copyright 2024 TIER IV, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "autoware/tensorrt_rtmdet/preprocess.hpp"
+
+#include <algorithm>
+#include <cstdio>
+#include <cstdlib>
+
+#define MIN(x, y) x < y ? x : y
+
+namespace autoware::tensorrt_rtmdet
+{
+constexpr size_t block = 512;
+constexpr size_t max_blocks_per_dim = 65535;
+constexpr int C = 3;
+
+dim3 cuda_gridsize(size_t n)
+{
+  size_t k = (n - 1) / block + 1;
+  size_t x = k;
+  size_t y = 1;
+  if (x > max_blocks_per_dim) {
+    x = ceil(sqrt(k));
+    y = (n - 1) / (x * block) + 1;
+  }
+  dim3 d;
+  d.x = x;
+  d.y = y;
+  d.z = 1;
+  return d;
+}
+
+__device__ double lerp1d(int a, int b, float w)
+{
+  return fma(w, (float)b, fma(-w, (float)a, (float)a));
+}
+
+__device__ float lerp2d(int f00, int f01, int f10, int f11, float centroid_h, float centroid_w)
+{
+  centroid_w = (1 + lroundf(centroid_w) - centroid_w) / 2;
+  centroid_h = (1 + lroundf(centroid_h) - centroid_h) / 2;
+
+  float r0, r1, r;
+  r0 = lerp1d(f00, f01, centroid_w);
+  r1 = lerp1d(f10, f11, centroid_w);
+
+  r = lerp1d(r0, r1, centroid_h);  //+ 0.00001
```

Comment:
`r0` and `r1` are float but args are int, was this intentional?

Reply:
I see, updated https://github.com/autowarefoundation/autoware.universe/pull/8165/commits/a7baba91e453f965dde364e3a7cabb69784ea549